### DISCUSSION TITLE
How to pass a request object as optional to a function in fastAPI?

### DISCUSSION THREAD

**Discussion by pradeepdev-1995** [fastapi/fastapi#6373] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI,Query
from starlette.requests import Request
from typing import Optional,
app = FastAPI()
@app.api_route("/endpoint/v1/{my_id}", methods=["GET", "POST", "DELETE"])
async def myFunction(request: Request, my_id: Optional[str] = None, request_type: Optional[str] = None):
#function body
                                                             ):
```

### Description

In the above code, I could able to handle the routing and could able to access the function - myFunction as
```
import requests
requests.get('http://localhost:7008/endpoint/v1/'),
```
Then I tried to call the function directly(without routing) as
```
myFunction()
```
But it shows the error like 
```
TypeError: myFunction() missing 1 required positional argument: 'request'
```
Now I have to doubts
**1 - How to call a fastAPI function directly (without routing)
2 - How to put the request object parameter as optional  in fastAPI function**

### Operating System

Linux

### FastAPI Version

fastapi==0.75.0

### Python Version

3.6.9

**Comment by jgould22** [comment, marked as answer]
FastAPI is a web framework and not really designed to have those functions be called as if they were library code.

If you need the functionality from that function extract it into another function/module/library and call that function from your route function and where ever else you might need it.

**Comment by pradeepdev-1995** [comment]
@jgould22  Thanks for the reply. So calling a route function directly (without routing, which is possible in flask) is not possible in fastAPI?

**Comment by pradeepdev-1995** [comment]
@jgould22  And any way to pass the **request** object parameter as optional in fastAPI function?

**Comment by chbndrhnns** [comment]
If you don't need the request, you could pass it as `None` or whatever fulfils your requirements. However, I agree with @jgould22 that these endpoint methods are not meant to be called directly.

**Comment by accelleon** [comment]
I'd reckon that if you need to call a route function anywhere inside of your backend it'd probably be best to split out the code that you're attempting to use into a new function and call that function from your endpoint. FastAPI is meant to define the public API for which web clients interact, not the internal interface that your code uses.

If its unit testing you're trying to accomplish, FastAPI has a `TestClient`.

**Comment by pradeepdev-1995** [comment]
@accelleon  Thanks for your reply

**Comment by pradeepdev-1995** [comment]
@chbndrhnns as you mentioned I tried to pass the request argument as optional(Default None) like below.
I tried two methods
1 - Method 1
```
@app.api_route("/endpoint/v1/{my_id}", methods=["GET", "POST", "DELETE"])
async def myFunction(request : Optional[Request] = None, my_id: Optional[str] = None, request_type: Optional[str] = None):
```
2- Method 2
```
import starlette
@app.api_route("/endpoint/v1/{my_id}", methods=["GET", "POST", "DELETE"])
async def myFunction(request : Optional[starlette.requests.Request] = None, my_id: Optional[str] = None, request_type: Optional[str] = None):
```
But both ways shows the error like this
```
fastapi.exceptions.FastAPIError: Invalid args for response field! Hint: check that typing.Union[starlette.requests.Request, NoneType] is a valid pydantic field type
```
How to resolve this?

**Comment by pradeepdev-1995** [comment]
@accelleon  I dont need to return the Request. 
I just want to put the request as optional. If so the function will invoke while routing, due to the presence of Request object argument. And the function will invoke when I am calling it without routing due to Request object is optional(because in direct function call, there is no request object) 
This is my scenario.

**Comment by accelleon** [comment]
@pradeepdev-1995
It seems as though the issue with `Optional[Request]` is that `Request` isn't a viable type for pydantic. Pydantic is pretty baked into FastAPI.

This is the actual exception that is being masked by FastAPI's
```
  File "pydantic/main.py", line 204, in pydantic.main.ModelMetaclass.__new__
  File "pydantic/fields.py", line 488, in pydantic.fields.ModelField.infer
  File "pydantic/fields.py", line 419, in pydantic.fields.ModelField.__init__
  File "pydantic/fields.py", line 539, in pydantic.fields.ModelField.prepare
  File "pydantic/fields.py", line 801, in pydantic.fields.ModelField.populate_validators
  File "pydantic/validators.py", line 723, in find_validators
RuntimeError: no validator found for <class 'starlette.requests.Request'>, see `arbitrary_types_allowed` in Config
```
FastAPI is bypassing that function call for dependencies and default parameters such as the type `Request`. You're gonna have to not declare the field as `Optional[]` and instead just pass it `None`. If you're using mypy or type checking in your IDE and want to rid yourself of the error you can use `type: ignore`:
```python
import starlette
@app.api_route("/endpoint/v1/{my_id}", methods=["GET", "POST", "DELETE"])
async def myFunction(request : starlette.requests.Request = None, my_id: Optional[str] = None, request_type: Optional[str] = None):  # type: ignore
```

**Comment by pradeepdev-1995** [comment]
@accelleon Thank you. The above solution solved my issue.
